tests/optics/focus-lens.test.tsx

```diff
@@ -0,0 +1,126 @@
+import { atom, Provider, useAtom } from 'jotai'
+import React from 'react'
+import * as rtl from '@testing-library/react'
+import { focusAtom } from '../../src/optics/focusAtom'
+
+const succ = (input: number) => input + 1
+
+it('basic derivation using focus works', async () => {
+  const bigAtom = atom({ a: 0 })
+  const aAtom = focusAtom(bigAtom, (optic) => optic.prop('a'))
+
+  const Counter: React.FC = () => {
+    const [count, setCount] = useAtom(aAtom)
+    const [bigAtomValue] = useAtom(bigAtom)
+    return (
+      <>
+        <div>bigAtom: {JSON.stringify(bigAtomValue)}</div>
+        <div>count: {count}</div>
+        <button onClick={() => setCount(succ)}>incr</button>
+        <button onClick={() => setCount(0)}>set zero</button>
+      </>
+    )
+  }
+
+  const { getByText, findByText } = rtl.render(
+    <Provider>
+      <Counter />
+    </Provider>
+  )
+
+  await findByText('count: 0')
+  await findByText('bigAtom: {"a":0}')
+
+  rtl.fireEvent.click(getByText('incr'))
+  await findByText('count: 1')
+  await findByText('bigAtom: {"a":1}')
+
+  rtl.fireEvent.click(getByText('incr'))
+  await findByText('count: 2')
+  await findByText('bigAtom: {"a":2}')
+
+  rtl.fireEvent.click(getByText('set zero'))
+  await findByText('count: 0')
+  await findByText('bigAtom: {"a":0}')
+})
+
+it('focus on an atom works', async () => {
+  const bigAtom = atom({ a: 0 })
+
+  const Counter: React.FC = () => {
+    const aAtom = focusAtom(bigAtom, (optic) => optic.prop('a'))
+    const [count, setCount] = useAtom(aAtom)
+    const [bigAtomValue] = useAtom(bigAtom)
+    return (
+      <>
+        <div>bigAtom: {JSON.stringify(bigAtomValue)}</div>
+        <div>count: {count}</div>
+        <button onClick={() => setCount(succ)}>button</button>
+      </>
+    )
+  }
+
+  const { getByText, findByText } = rtl.render(
+    <Provider>
+      <Counter />
+    </Provider>
+  )
+
+  await findByText('count: 0')
+  await findByText('bigAtom: {"a":0}')
+
+  rtl.fireEvent.click(getByText('button'))
+  await findByText('count: 1')
+  await findByText('bigAtom: {"a":1}')
+})
+
+it('double-focus on an atom works', async () => {
+  const bigAtom = atom({ a: { b: 0 } })
+
+  const Counter: React.FC = () => {
+    const [bigAtomValue, setBigAtom] = useAtom(bigAtom)
+    const atomA = focusAtom(bigAtom, (optic) => optic.prop('a'))
+    const atomB = focusAtom(atomA, (optic) => optic.prop('b'))
```

Comment:
Should we use useMemo here? for tests could be a good reference.

Reply:
Or, can we move these out of Counter?

Reply:
fixed